qemu/tests/fio_perf.py

```diff
@@ -95,38 +112,52 @@ def run(test, params, env):
     :param env: Dictionary with test environment
     """
 
-    def fio_thread(session, run_cmd, cmd_timeout):
+    def fio_thread():
         """
         run fio command in guest
-
-        :param session: VM session
-        :param run_cmd: fio command
-        :param cmd_timeout: Timeout in seconds
         """
         session.cmd_status(run_cmd, cmd_timeout)
 
-    def _pin_vm_threads(vm, node):
+    def _pin_vm_threads(node):
         """
         pin guest vcpu and vhost threads to cpus of a numa node repectively
 
-        :param vm: VM object
         :param node: which numa node to pin
         """
         if node:
             if not isinstance(node, utils_misc.NumaNode):
                 node = utils_misc.NumaNode(int(node))
             utils_test.qemu.pin_vm_threads(vm, node)
 
+    def fio_install(tarball):
+        """
+        check whether fio is installed in guest, if no, install it, if yes, do nothing.
+
+        :param tarball: fio tar package
+        """
+        if session.cmd_status(check_install_fio):
+            tarball = os.path.join(data_dir.get_deps_dir(), tarball)
+            if os_type == "linux":
+                vm.copy_files_to(tarball, "/tmp")
+                session.cmd("cd /tmp/ && tar -zxvf /tmp/fio-2.2.9.tar.gz", cmd_timeout)
+                session.cmd("cd /tmp/fio-2.2.9 && %s" % compile_cmd)
+            elif os_type == "windows":
+                session.cmd("md C:\\fio-2.0.15-x64")
+                vm.copy_files_to(tarball, "C:\\fio-2.0.15-x64")
+
     # login virtual machine
     vm = env.get_vm(params["main_vm"])
     vm.verify_alive()
     login_timeout = int(params.get("login_timeout", 360))
     session = vm.wait_for_login(timeout=login_timeout)
     process.system_output("numactl --hardware")
     process.system_output("numactl --show")
-    _pin_vm_threads(vm, params.get("numa_node"))
+    _pin_vm_threads(params.get("numa_node"))
 
     # get parameter from dictionary
+    check_install_fio = params["check_install_fio"]
+    tarball = params["tarball"]
+    compile_cmd = params.get("compile_cmd")
```

Comment:
Please use params["compile_cmd"] here, otherwise you will not catch the error when missing this parameter, or you should set a default value here.